contracts/exchangeIssuance/ExchangeIssuance.sol

```diff
@@ -0,0 +1,587 @@
+import { IUniswapV2Factory } from "@uniswap/v2-core/contracts/interfaces/IUniswapV2Factory.sol";
+
+import { PreciseUnitMath } from "../lib/PreciseUnitMath.sol";
+import { UniswapV2Library } from "../../external/contracts/UniswapV2Library.sol";
+import { SushiswapV2Library } from "../../external/contracts/SushiswapV2Library.sol";
+import { ISetToken } from "../interfaces/ISetToken.sol";
+import { IBasicIssuanceModule } from "../interfaces/IBasicIssuanceModule.sol";
+import { IWETH } from "../interfaces/IWETH.sol";
+
+/**
+ * @title ExchangeIssuance
+ * @author Index Coop
+ *
+ * Contract for minting and redeeming any Set token using
+ * ETH or an ERC20 as the paying/receiving currency. All swaps are done using the best price
+ * found on Uniswap or Sushiswap.
+ *
+ */
+contract ExchangeIssuance is ReentrancyGuard {
+
+    using SafeMath for uint256;
+
+    /* ============ constants ============ */
+    
+    uint256 constant private MAX_UINT96 = 2 ** 96 - 1;
+    
+    /* ============ State Variables ============ */
+
+    IUniswapV2Router02 private uniRouter;
+    address private uniFactory;
+    IUniswapV2Router02 private sushiRouter;
+    address private sushiFactory;
+
+    IBasicIssuanceModule private basicIssuanceModule;
+    address private WETH;
+
+    /* ============ Events ============ */
+
+    event ExchangeIssue(
+        address indexed _recipient,
+        address indexed _setToken,
+        address indexed _inputToken,
+        uint256 _amountSetToken
+    );
+    event ExchangeRedeem(
+        address indexed _recipient,
+        address indexed _setToken,
+        address indexed _outputToken,
+        uint256 _amountSetToken,
+        uint256 _amountOut
+    );
+
+    /* ============ Constructor ============ */
+
+    constructor(
+        address _uniFactory,
+        IUniswapV2Router02 _uniRouter, 
+        address _sushiFactory, 
+        IUniswapV2Router02 _sushiRouter, 
+        IBasicIssuanceModule _basicIssuanceModule
+    )
+        public
+    {
+        uniFactory = _uniFactory;
+        uniRouter = _uniRouter;
+
+        sushiFactory = _sushiFactory;
+        sushiRouter = _sushiRouter;
+
+        WETH = uniRouter.WETH();
+        basicIssuanceModule = _basicIssuanceModule;
+        IERC20(WETH).approve(address(uniRouter), PreciseUnitMath.maxUint256());
+        IERC20(WETH).approve(address(sushiRouter), PreciseUnitMath.maxUint256());
+    }
+
+    /* ============ External Functions ============ */
+
+    /**
+     * Runs all the necessary approval functions required before issuing
+     * or redeeming a set token. This function only need to be called before the first time
+     * this smart contract is used on any particular set token, or when a new token is added
+     * to the set token during a rebalance.
+     *
+     * @param _setToken    Address of the set token being initialized
+     */
+    function initApprovals(ISetToken _setToken) external {
```

Comment:
Based on what the function is doing I’d suggest we’d:
- Pass in a list of tokens to enable it, this should not be coupled to a SetToken

Reply:
Added 
- `approveToken(IERC20 _token)`
- `approveTokens(IERC20[] tokens)` 
- Renamed `initApprovals` to `approveSetToken(ISetToken _setToken)`. 

I think, we can keep the `approveSetToken()` (which is coupled to a SetToken) cause this function will be used the most.
And it is cleaner from the frontend perspecitve, to just call `approveSetToken()` rather than, forming a list of tokens from the `setToken.getPostions()`, and then calling `approveTokens()`.